File: src/AreaExporterService.py
import json
import gmsh
from typing import Dict, List
import numpy as np
class AreaExporterService:
    _EMPTY_NAME_CASE = ""
    computedAreas:Dict[str,List]
    geometry: Dict
    def __init__(self):
        self.computedAreas = {
            "geometries": []
        }
    
    def addComputedArea(self, geometry:str, label:str, area:float):
        geometry:Dict ={
            "geometry": geometry,
            "label": label,
            "area": round(area,6),
        }
        self.computedAreas['geometries'].append(geometry)

    def addPhysicalModelForConductors(self, mappedElements:Dict[str,str]):
        physicalGroups = gmsh.model.getPhysicalGroups(1)
        for physicalGroup in physicalGroups:
            entityTags = gmsh.model.getEntitiesForPhysicalGroup(*physicalGroup)
            geometryName = gmsh.model.getPhysicalName(*physicalGroup)
            if geometryName.startswith("Conductor_"):
                label = ''
                for key, geometry in mappedElements.items():
                    if geometry == geometryName:
                        label = key
                        break
                
                    assert len(entityTags) == 1
                    loop = gmsh.model.geo.addCurveLoop(entityTags)
                    surface = gmsh.model.geo.addPlaneSurface([loop])

                    gmsh.model.geo.synchronize()

                    area = gmsh.model.occ.getMass(2, surface)
                    gmsh.model.occ.remove([(2, surface)])
                    if geometryName != AreaExporterService._EMPTY_NAME_CASE:
                        self.addComputedArea(geometryName, label, area)

    def exportToJson(self, exportFileName:str):
        with open(exportFileName + ".areas.json", 'w') as f:
            json.dump(self.computedAreas, f, indent=3)

File: step2gmsh/src/AreaExporterService.cpp
#include "AreaExporterService.h"

#include <cmath>
#include <fstream>
#include <set>

#include <gmsh.h>

namespace step2gmsh {

AreaExporterService::AreaExporterService() {
    computedAreas["geometries"] = nlohmann::json::array();
}

void AreaExporterService::addComputedArea(const std::string& geometry,
                                          const std::string& label,
                                          double area)
{
    double rounded = std::round(area * 1e6) / 1e6;
    computedAreas["geometries"].push_back({
        {"geometry", geometry},
        {"label",    label},
        {"area",     rounded}
    });
}

void AreaExporterService::addPhysicalModelForConductors(
    const std::map<std::string, std::string>& mappedElements)
{
    gmsh::vectorpair physicalGroups;
    gmsh::model::getPhysicalGroups(physicalGroups, 1);

    gmsh::vectorpair allSurfaces;
    gmsh::model::getEntities(allSurfaces, 2);

    for (const auto& [pgDim, pgTag] : physicalGroups) {
        std::string geometryName;
        gmsh::model::getPhysicalName(pgDim, pgTag, geometryName);

        if (geometryName.rfind("Conductor_", 0) != 0) continue;

        std::vector<int> entityTags;
        gmsh::model::getEntitiesForPhysicalGroup(pgDim, pgTag, entityTags);
        std::set<int> entityTagSet(entityTags.begin(), entityTags.end());

        // Find the label associated with this geometry name
        std::string label;
        for (const auto& [key, geom] : mappedElements) {
            if (geom == geometryName) {
                label = key;
                break;
            }
        }

        // Find the surface whose boundary matches the curve tags
        for (const auto& surface : allSurfaces) {
            gmsh::vectorpair boundary;
            gmsh::model::getBoundary({surface}, boundary, true, false, false);
            std::set<int> boundaryTags;
            for (const auto& [dim, tag] : boundary) {
                boundaryTags.insert(tag);
            }
            if (entityTagSet == boundaryTags) {
                double area = 0.0;
                gmsh::model::occ::getMass(2, surface.second, area);
                addComputedArea(geometryName, label, area);
                break;
            }
        }
    }
}

void AreaExporterService::exportToJson(const std::string& exportFileName) const {
    std::ofstream f(exportFileName + ".areas.json");
    f << computedAreas.dump(3);
}

} // namespace step2gmsh
